src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipUnArchiver.java

```diff
@@ -245,30 +235,49 @@ protected void execute( final String path, final File outputDirectory )
 
                 if ( ze.getName().startsWith( path ) )
                 {
-                    in = zipFile.getInputStream( ze );
-
-                    extractFileIfIncluded( getSourceFile(), outputDirectory, in,
-                                           ze.getName(), new Date( ze.getTime() ), ze.isDirectory(),
-                                           ze.getUnixMode() != 0 ? ze.getUnixMode() : null,
-                                           resolveSymlink( zipFile, ze ), getFileMappers() );
-
-                    in.close();
-                    in = null;
+                    try(InputStream in = zipFile.getInputStream( ze )) {
+                        if (remainingSpace != null)
+                        {
+                            File file = extractFile(getSourceFile(), outputDirectory, in, remainingSpace,
+                                    ze.getName(), new Date(ze.getTime()), ze.isDirectory(),
+                                    ze.getUnixMode() != 0 ? ze.getUnixMode() : null,
+                                    resolveSymlink(zipFile, ze), getFileMappers());
+                            if (file != null)
+                            {
+                                remainingSpace -= file.length();
```

Comment:
Is `file.length()` reliable way to get the file size. What I mean is (quoting the `File` class API docs):
>Where it is required to distinguish an I/O exception from the case that `0L` is returned, or where several attributes of the same file are required at the same time, then the `java.nio.file.Files#readAttributes` method may be used.